tsdb/store.go

```diff
@@ -1709,67 +1709,87 @@ func (s *Store) TagValues(ctx context.Context, auth query.Authorizer, shardIDs [
 			switch e.Op {
 			case influxql.EQ, influxql.NEQ, influxql.EQREGEX, influxql.NEQREGEX:
 				tag, ok := e.LHS.(*influxql.VarRef)
-				if !ok || tag.Val != "_name" {
+				if !ok || tag.Val != measurementKey {
 					return nil
 				}
 			}
 		}
 		return e
 	}), nil)
 
-	filterExpr := influxql.CloneExpr(cond)
-	filterExpr = influxql.Reduce(influxql.RewriteExpr(filterExpr, func(e influxql.Expr) influxql.Expr {
+	is, err := s.indexSetFromShards(ctx, shardIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	names, err := is.MeasurementNamesByExpr(nil, measurementExpr)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.tagValues(ctx, names, is, auth, cond)
+}
+
+// TagValuesByPredicate gets a slice of tag values, provided a flux predicate
+// that has been parsed into an influxql.Expr. It differs from TagValues in that
+// it works with flux predicates.
+func (s *Store) TagValuesByPredicate(ctx context.Context, auth query.Authorizer, shardIDs []uint64, cond influxql.Expr) ([]TagValues, error) {
+	is, err := s.indexSetFromShards(ctx, shardIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	measurementExpr := influxql.CloneExpr(cond)
+	measurementExpr = influxql.Reduce(influxql.RewriteExpr(measurementExpr, func(e influxql.Expr) influxql.Expr {
 		switch e := e.(type) {
 		case *influxql.BinaryExpr:
 			switch e.Op {
 			case influxql.EQ, influxql.NEQ, influxql.EQREGEX, influxql.NEQREGEX:
 				tag, ok := e.LHS.(*influxql.VarRef)
-				if !ok || influxql.IsSystemName(tag.Val) {
+				// Drop all system names except "_name" so that regex and negation
+				// operators work. The primary effect of this is dropping the "_tagKey"
+				// part of the expression for retrieving measurement names.
+				if !ok || (tag.Val != measurementKey && influxql.IsSystemName(tag.Val)) {
 					return nil
 				}
 			}
 		}
 		return e
 	}), nil)
 
-	// Build index set to work on.
-	is := IndexSet{Indexes: make([]Index, 0, len(shardIDs))}
-	s.mu.RLock()
-	for _, sid := range shardIDs {
-		shard, ok := s.shards[sid]
-		if !ok {
-			continue
-		}
+	names, err := is.MeasurementNamesByPredicate(nil, measurementExpr)
+	if err != nil {
+		return nil, err
+	}
 
-		if is.SeriesFile == nil {
-			sfile, err := shard.SeriesFile()
-			if err != nil {
-				s.mu.RUnlock()
-				return nil, err
-			}
-			is.SeriesFile = sfile
-		}
+	// The list of names must be sorted, since MeasurementNamesByPredicate does
+	// not sort.
+	sort.Slice(names, func(i, j int) bool {
+		return bytes.Compare(names[i], names[j]) < 0
+	})
 
-		index, err := shard.Index()
-		if err != nil {
-			s.mu.RUnlock()
-			return nil, err
-		}
+	return s.tagValues(ctx, names, is, auth, cond)
+}
 
-		is.Indexes = append(is.Indexes, index)
-	}
-	s.mu.RUnlock()
+func (s *Store) tagValues(ctx context.Context, names [][]byte, is IndexSet, auth query.Authorizer, cond influxql.Expr) ([]TagValues, error) {
+	filterExpr := influxql.CloneExpr(cond)
```

Comment:
Everything in the `tagValues` method is nearly identical to the previous `TagValues` method, minus getting the measurement names and created the `IndexSet`.